src/search.rs

```diff
@@ -214,6 +214,47 @@ struct GoalFrame {
     cut_level: usize,
 }
 
+/// A single assignment of variables to terms.
+///
+/// The variables returned in this struture correspond to the goals via [ordinals](crate::ast::Var::ord).
+/// i.e. the assignment for a variable `x: Var` is stored at index
+/// `x.ord()`.
+///
+/// Names of variables are not stored here. They can be resolved using the matching Scope.
+///
+/// Note that even in valid solutions, not all variables need to have an assignment. For
+/// example, given the rule
+///
+/// ```prolog
+/// anything(X).
+/// ```
+///
+/// the query `anything(X)` will be true for any `X`, hence even a valid solution won't have an
+/// assignment for this variable.
+#[derive(PartialEq, Debug, Clone)]
+pub struct Solution(pub Vec<Option<ast::Term>>);
+
+impl Solution {
+    /// Returns a reference to the vector containing all assignments of terms to variables.
+    /// The vector's indices correspond to ordinals of the variables.
+    pub fn vars(&self) -> &Vec<Option<ast::Term>> {
+        &self.0
+    }
+
+    /// Iterates over all variables and their assigned terms.
+    pub fn iter_vars(&self) -> impl Iterator<Item = (Var, &Option<ast::Term>)> {
```

Comment:
For consistency, perhaps `get` and `iter_vars` should both either return `&Option<Term>` or `Option<&Term>`, rather than a mix.
I'm not sure what's better here. Usually a `Option<&T>` is a bit more ergonomic to work with, but you can get a `Option<&T>` from `&Option<T>` via `as_ref()`, but you cannot go the other way around.

That said, I'd still go with `Option<&Term>` consistently here, because the only promise a `Solution` makes is that for each goal variable, either there's a term, or there's not. But providing `&Option<Term>` promises a concrete implementation detail (namely that each variable is explicitly associated with an `Option`.
That way, it wouldn't be possible to transparently swap out the backing storage to e.g. a `HashMap` that only contains variables that have a value. (Not that such a change would be useful here, but it could be in the future).

Reply:
Fixed.

I like the way you review, your feedback is about things that are actual code smells!

Reply:
Thanks :)